src/iosrv/sched.c

```diff
@@ -0,0 +1,539 @@
+/**
+ * (C) Copyright 2016-2019 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+
+#define D_LOGFAC       DD_FAC(server)
+
+#include <abt.h>
+#include <daos/common.h>
+#include <daos_errno.h>
+#include "srv_internal.h"
+
+/*
+ * A schedule cycle consists of three stages:
+ * 1. Starting with a network poll ULT, number of ULTs to be executed in this
+ *    cycle is queried by ABT_pool_get_size() for each non-poll ABT pool;
+ * 2. Executing all the non-poll ULTs;
+ * 3. Ending with a NVMe poll ULT;
+ *
+ * Extra network & NVMe poll ULTs could be scheduled in executing stage
+ * according to network/NVMe poll age and request/IO statistics.
+ */
+struct sched_cycle {
+	uint32_t	sc_ults_cnt[DSS_POOL_CNT];
+	uint32_t	sc_ults_tot;
+	uint32_t	sc_age_net_bound[2];
+	uint32_t	sc_age_nvme_bound[2];
+	uint32_t	sc_age_net;
+	uint32_t	sc_age_nvme;
+	unsigned int	sc_new_cycle:1,
+			sc_cycle_started:1;
+};
+
+struct sched_data {
+	struct sched_cycle	 sd_cycle;
+	struct dss_xstream	*sd_dx;
+	uint32_t		 sd_event_freq;
+};
+
+static unsigned int sched_throttle[DSS_POOL_CNT] = {
+	0,	/* DSS_POOL_NET_POLL */
+	0,	/* DSS_POOL_NVME_POLL */
+	0,	/* DSS_POOL_IO */
+	30,	/* DSS_POOL_REBUILD */
+	0,	/* DSS_POOL_AGGREGATE */
+	0,	/* DSS_POOL_GC */
```

Comment:
just wondering any advantage to separate the aggregate and GC pool? (can use one pool for that both?)

Reply:
Yeah, these two jobs are for space reclaiming, so they likely always have same priority, probably we can put them into same ABT pool.

But keeping them in separate pools could have the advantage of flexibility, for instance, sometime we may want to throttle aggregation, but don't want to throttle GC (it usually start working when a large container is destroyed).